Handle SIGPIPE/BrokenPipeError in pytest's CLI

bluetech opened the pull request:
Running `pytest | head -1` and similar causes an annoying error to be
printed to stderr:

    Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
    BrokenPipeError: [Errno 32] Broken pipe

(or possibly even a propagating exception in older/other Python versions).

The standard UNIX behavior is to handle the SIGPIPE silently. To
recommended method to do this in Python is described here:
https://docs.python.org/3/library/signal.html#note-on-sigpipe

It is not appropriate to apply this recommendation to `pytest.main()`,
which is used programmatically for in-process runs. Hence, change
pytest's entrypoint to a new `pytest.console_main()` function, to be
used exclusively by pytest's CLI, and add the SIGPIPE code there.

Fixes #4375.

nicoddemus commented on src/_pytest/config/__init__.py:
```diff
@@ -137,6 +137,24 @@ def main(args=None, plugins=None) -> Union[int, ExitCode]:
         return ExitCode.USAGE_ERROR
 
 
+def console_main() -> int:
```
Can we test this somehow? Perhaps with `testdir.runpytest_subprocess`?

bluetech replied:
Let me try...

RonnyPfannschmidt commented:
It's a improvement

bluetech commented on setup.cfg:
```diff
@@ -43,8 +43,8 @@ python_requires = >=3.5
 console_scripts =
-	pytest=pytest:main
-	py.test=pytest:main
+	pytest=pytest:console_main
```
It feels wrong to me to use a private function as the endpoint. I guess I just want the entry point to `pytest` to be in `pytest` :smile: 

Since I don't expect this would ever need to be changed (the signature being strictly specified), do you mind if I keep it as is?

BTW, I was originally thinking the wrapper should just be straight in `pytest/__main__.py`, but decided against it since we don't want any functional code there.

nicoddemus replied:
> Since I don't expect this would ever need to be changed (the signature being strictly specified)

I'm not worried about the signature changing, is just that being public API we need to keep it around even if we don't need it anymore in the future for example. But OK, I don't think this is a big deal.

> do you mind if I keep it as is?

Sure, I don't feel strongly about it. 👍

bluetech commented:
- Added a test (it seems Windows behaves differently in a way that can't be handled really, so I skipped it there -- I don't have a Windows to test with unfortunately).
- Added changelog.

bluetech commented:
I'm not sure why the `BrokenPipeError` isn't covered in CI -- the tests passes, and it is covered when I run it locally, and the test fails without the changes. Maybe something different about the environment.
